motion_computation/src/bsm_to_external_object_convertor.cpp

```diff
@@ -0,0 +1,348 @@
+#include <carma_perception_msgs/msg/external_object.hpp>
+#include <carma_v2x_msgs/msg/psm.hpp>
+#include <motion_computation/impl/bsm_to_external_object_helpers.hpp>
+#include <motion_computation/message_conversions.hpp>
+
+namespace motion_computation {
+
+namespace conversion {
+
+void convert(const carma_v2x_msgs::msg::BSM& in_msg, carma_perception_msgs::msg::ExternalObject& out_msg,
+             const std::string& map_frame_id, double pred_period, double pred_step_size,
+             const lanelet::projection::LocalFrameProjector& map_projector, tf2::Quaternion ned_in_map_rotation) {
+
+  out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::BSM_ID_PRESENCE_VECTOR;
+
+  if (true) //TODO:: check how to determine an object is dynamic???
+  {
+    /////// Dynamic Object /////////
+    out_msg.dynamic_obj = true;
+    out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::DYNAMIC_OBJ_PRESENCE;
+
+  }
+
+  /////// Object Id /////////
+  // Generate a unique object id from the psm id
+  out_msg.id = 0;
+  for (int i = in_msg.core_data.id.size() - 1; i >= 0; i--) {  // using signed iterator to handle empty case
+    // each byte of the bsm id gets placed in one byte of the object id.
+    // This should result in very large numbers which will be unlikely to
+    // conflict with standard detections
+    out_msg.id |= in_msg.core_data.id[i] << (8 * i);
+  }
+  out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::ID_PRESENCE_VECTOR;
+
+
+  /////// BSM Id /////////
+  out_msg.bsm_id = in_msg.core_data.id;
+  out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::BSM_ID_PRESENCE_VECTOR;
+
+
+  /////// Timestamp /////////
+  // Compute the timestamp
+  out_msg.header.stamp = in_msg.header.stamp;
+  out_msg.header.frame_id = map_frame_id;
+
+  /////// Object Type and Size /////////
+  
+  if (in_msg.core_data.size.presence_vector & carma_v2x_msgs::msg::VehicleSize::VEHICLE_LENGTH_AVAILABLE)
+  {
+    out_msg.size.x = in_msg.core_data.size.vehicle_length;
```

Comment:
this should be divided by 2 since ExternalObject size is half of each dimension. Same for width